types/src/proof/definition.rs

```diff
@@ -45,16 +57,83 @@ impl AccumulatorProof {
             assert_ne!(*first_sibling, *ACCUMULATOR_PLACEHOLDER_HASH);
         }
 
-        AccumulatorProof { siblings }
+        AccumulatorProof {
+            siblings,
+            phantom: PhantomData,
+        }
     }
 
     /// Returns the list of siblings in this proof.
     pub fn siblings(&self) -> &[HashValue] {
         &self.siblings
     }
+
+    /// Verifies an element whose hash is `element_hash` and version is `element_version` exists in
+    /// the accumulator whose root hash is `expected_root_hash` using the provided proof.
+    pub fn verify(
+        &self,
+        expected_root_hash: HashValue,
+        element_hash: HashValue,
+        element_index: u64,
+    ) -> Result<()> {
+        ensure!(
+            self.siblings.len() <= MAX_ACCUMULATOR_PROOF_DEPTH,
+            "Accumulator proof has more than {} ({}) siblings.",
+            MAX_ACCUMULATOR_PROOF_DEPTH,
+            self.siblings.len()
+        );
+
+        let actual_root_hash = self
+            .siblings
+            .iter()
+            .rev()
+            .fold(
+                (element_hash, element_index),
+                // `index` denotes the index of the ancestor of the element at the current level.
+                |(hash, index), sibling_hash| {
+                    (
+                        if index % 2 == 0 {
+                            // the current node is a left child.
+                            MerkleTreeInternalNode::<H>::new(hash, *sibling_hash).hash()
+                        } else {
+                            // the current node is a right child.
+                            MerkleTreeInternalNode::<H>::new(*sibling_hash, hash).hash()
+                        },
+                        // The index of the parent at its level.
+                        index / 2,
+                    )
+                },
+            )
+            .0;
+        ensure!(
+            actual_root_hash == expected_root_hash,
+            "Root hashes do not match. Actual root hash: {:x}. Expected root hash: {:x}.",
+            actual_root_hash,
+            expected_root_hash
+        );
+
+        Ok(())
+    }
+}
+
+impl<H> std::fmt::Debug for AccumulatorProof<H> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "AccumulatorProof {{ siblings: {:?} }}", self.siblings)
+    }
 }
 
-impl TryFrom<crate::proto::types::AccumulatorProof> for AccumulatorProof {
+impl<H> PartialEq for AccumulatorProof<H> {
```

Comment:
Because `H` doesn't implement `Eq` and `PartialEq`.

Reply:
🙈

Reply:
Ok, then why not derive `Eq`?

Reply:
You mean

```
diff --git a/types/src/proof/definition.rs b/types/src/proof/definition.rs
index eddf1270..e03a832c 100644
--- a/types/src/proof/definition.rs
+++ b/types/src/proof/definition.rs
@@ -28,7 +28,7 @@ use std::marker::PhantomData;
 /// A proof that can be used authenticate an element in an accumulator given trusted root hash. For
 /// example, both `LedgerInfoToTransactionInfoProof` and `TransactionInfoToEventProof` can be
 /// constructed on top of this structure.
-#[derive(Clone)]
+#[derive(Clone, Eq)]
 pub struct AccumulatorProof<H> {
     /// All siblings in this proof, including the default ones. Siblings near the root are at the
     /// beginning of the vector.
@@ -128,8 +128,6 @@ impl<H> PartialEq for AccumulatorProof<H> {
     }
 }

-impl<H> Eq for AccumulatorProof<H> {}
-
 impl<H> TryFrom<crate::proto::types::AccumulatorProof> for AccumulatorProof<H>
 where
     H: CryptoHasher,
```

? Getting same error:

```
+ cargo build
   Compiling libra-types v0.1.0 (/Users/wqfish/github/libra/types)
error[E0277]: the trait bound `crypto::hash::TransactionAccumulatorHasher: std::cmp::Eq` is not satisfied
   --> types/src/proof/definition.rs:382:5
    |
382 |     ledger_info_to_transaction_info_proof: TransactionAccumulatorProof,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::cmp::Eq` is not implemented for `crypto::hash::TransactionAccumulatorHasher`
    |
    = note: required because of the requirements on the impl of `std::cmp::Eq` for `proof::definition::AccumulatorProof<crypto::hash::TransactionAccumulatorHasher>`
    = note: required by `std::cmp::AssertParamIsEq`

error[E0277]: the trait bound `crypto::hash::TransactionAccumulatorHasher: std::cmp::Eq` is not satisfied
```

Reply:
weird.. 

@bors-libra delegate

Reply:
I think it's just the procedural macro implementation enforces that any generic parameter must implement `PartialEq` and `Eq`. Something like [this](https://github.com/dtolnay/syn/blob/2218cd20d8444f925aafbfd166562dd122106f39/examples/heapsize/heapsize_derive/src/lib.rs#L16).